Sources/StreamChat/Utils/StreamCollection.swift

```diff
@@ -0,0 +1,41 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+/// A collection wrapping the base collection which allows base collections of different types.
+///
+/// - Note: The type of the base collection can change in the future.
+public struct StreamCollection<Element>: RandomAccessCollection {
+    public typealias Index = Int
+
+    private let _endIndex: () -> Index
+    private let _position: (Index) -> Element
+    private let _startIndex: () -> Index
```

Comment:
Using closures allows hiding the type of the base collection since for now we want to use `LazyCachedMapCollection`, later `Array`.

Reply:
I'm not sure I understand the benefit of using closures here. Do you mind elaborating?